File: LeakyDocObj-CDemo/Document.m
#import "Document.h"
//#import <objc/runtime.h>

void objc_release(id object);

static NSInteger nExtraReleases = 3;

@interface Document ()

@end

@implementation Document

- (instancetype)init {
    self = [super init];

    if (self) {
        // Add your subclass-specific initialization here.
    }
    return self;
}

+ (BOOL)autosavesInPlace {
    return YES;
}


- (void)makeWindowControllers {
    // Override to return the Storyboard file name of the document.
    [self addWindowController:[[NSStoryboard storyboardWithName:@"Main" bundle:nil] instantiateControllerWithIdentifier:@"Document Window Controller"]];
}


- (NSData *)dataOfType:(NSString *)typeName error:(NSError **)outError {
    return [NSData data];
}


- (BOOL)readFromData:(NSData *)data ofType:(NSString *)typeName error:(NSError **)outError {
    return YES;
}

- (void)dealloc {
    NSLog(@"Dealloc doc %p", self) ;
}

- (NSError*)willPresentError:(NSError *)error {
    if (
        [[error domain] isEqualToString:NSCocoaErrorDomain]
        &&
        ([error code] == 67001)
        ) {

        NSLog(@"Doing %ld extra releases to %p", nExtraReleases, self) ;
        for (NSInteger i=0; i<nExtraReleases; i++) {
            objc_release(self);
        }

        nExtraReleases++;
    }

    return [super willPresentError:error];
}


@end

File: LeakyDocSwiftDemo/Document.swift
import Cocoa

var nExtraReleases = 3

class Document: NSDocument {

    override init() {
        super.init()
        // Add your subclass-specific initialization here.
    }

    override class var autosavesInPlace: Bool {
        return true
    }

    override func makeWindowControllers() {
        // Returns the Storyboard that contains your Document window.
        let storyboard = NSStoryboard(name: NSStoryboard.Name("Main"), bundle: nil)
        let windowController = storyboard.instantiateController(withIdentifier: NSStoryboard.SceneIdentifier("Document Window Controller")) as! NSWindowController
        self.addWindowController(windowController)
    }

    override func data(ofType typeName: String) throws -> Data {
        let someData = UnsafeRawPointer(bitPattern: 1)!
        return Data(bytes: someData, count: 0)
    }

    override func read(from data: Data, ofType typeName: String) throws {
    }

    deinit {
        Swift.print("Dealloc doc \(self)")
    }

    override func willPresentError(_ error: Error) -> Error {
        Swift.print("Presenting!")
        if (error._domain == NSCocoaErrorDomain) && (error._code == 67001) {
            Swift.print("Doing \(nExtraReleases) extra releases to \(self)")
            let unmanagedSelf = Unmanaged.passUnretained(self)
            for _ in 0..<nExtraReleases {
                unmanagedSelf.release()
            }
            nExtraReleases += 1
        }
        return super.willPresentError(error)
    }

}

